File: cpp/0054_Spiral_Matrix.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        int top = 0, bottom = m - 1, left = 0, right = n - 1;
        int d = 1;
        vector<int> ans;

        while (bottom >= top && right >= left) {
            if (d == 1) {
                for (int c = left; c <= right; c++) ans.push_back(matrix[top][c]);
                top++;
            } else if (d == 2) {
                for (int r = top; r <= bottom; r++) ans.push_back(matrix[r][right]);
                right--;
            } else if (d == 3) {
                for (int c = right; c >= left; c--) ans.push_back(matrix[bottom][c]);
                bottom--;
            } else {
                for (int r = bottom; r >= top; r--) ans.push_back(matrix[r][left]);
                left++;
            }
            d = (d + 1) % 4;
        }

        return ans;
    }
};

File: python/0054_Spiral_Matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        top, bottom = 0, m - 1
        left, right = 0, n - 1
        d = 1
        ans = []

        while bottom >= top and right >= left:
            if d == 1:
                for c in range(left, right + 1):
                    ans.append(matrix[top][c])
                top += 1
            elif d == 2:
                for r in range(top, bottom + 1):
                    ans.append(matrix[r][right])
                right -= 1
            elif d == 3:
                for c in range(right, left - 1, -1):
                    ans.append(matrix[bottom][c])
                bottom -= 1
            else:
                for r in range(bottom, top - 1, -1):
                    ans.append(matrix[r][left])
                left += 1
            d = (d + 1) % 4
        
        return ans